Source/Pawnmorphs/Esoteria/UserInterface/FilterTreeBox.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Unix.Native;
+using Pawnmorph.UserInterface.TreeBox;
+using Pawnmorph.Utilities.Collections;
+using TMPro;
+using UnityEngine;
+using Verse;
+
+namespace Pawnmorph.UserInterface
+{
+	/// <summary>
+	/// Searchable list box.
+	/// </summary>
+	internal class FilterTreeBox
+	{
+		private Vector2 _scrollPosition = new Vector2(0, 0);
+		private string _currentFilter = string.Empty;
+		private List<TreeNode_FilterBox> _roots;
+		private List<TreeNode_FilterBox> _items;
+
+		public FilterTreeBox(List<TreeNode_FilterBox> roots)
+		{
+			_roots = roots.ToList();
+			_roots.Reverse();
+			_items = new List<TreeNode_FilterBox>();
+			UpdateFilter();
+		}
+
+
+		private void UpdateFilter()
+		{
+			_items.Clear();
+			// Reset tree.
+			if (_currentFilter.Length == 0)
+			{
+				for (int i = _roots.Count - 1; i >= 0; i--)
+				{
+					_roots[i].SetVisibility(true, true);
+					_roots[i].GetVisibleNodes(_items);
+				}
+			}
+			else
+			{
+				for (int i = _roots.Count - 1; i >= 0; i--)
+				{
+					_roots[i].SetVisibility(false, false);
+					TreeSearch(_roots[i]);
+					_roots[i].GetVisibleNodes(_items);
+				}
+			}
+		}
+
+		private void UpdateVisibleNodes()
+		{
+			_items.Clear();
+			for (int i = _roots.Count - 1; i >= 0; i--)
+			{
+				_roots[i].GetVisibleNodes(_items);
+			}
+		}
+
+		private void TreeSearch(TreeNode_FilterBox node)
+		{
+			if (node.Label.ToLower().Contains(_currentFilter))
```

Comment:
Is `_currentFilter` forced lowercase anywhere?  If not, it should have a .ToLower() here as well

Reply:
Yes, in line 102 when the text is retrieved from the textbox.
It is somewhat obscure though, however needed to ensure the comparison between old and new text detects changes properly. (line 106)